File: benchmarks/executable/restaurant_uint_wrapping_shift_left.c
// C23 reference for the Restaurant UInt wrapping-shift-left workload.

#include <inttypes.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>

#ifdef _WIN32
#include <fcntl.h>
#include <io.h>
#endif

static volatile uint64_t runtime_value = UINT64_MAX;
static volatile uint64_t runtime_count = UINT64_C(1);

static bool wrapping_shift_left_u64(uint64_t value, uint64_t count,
                                    uint64_t *result) {
  const bool valid_count = count < UINT64_C(64);
  if (!valid_count) return false;
  *result = value << count;
  return true;
}

int main(void) {
#ifdef _WIN32
  if (_setmode(_fileno(stdout), _O_BINARY) == -1) return 1;
#endif
  uint64_t wrapped = UINT64_C(0);
  if (!wrapping_shift_left_u64(runtime_value, runtime_count, &wrapped)) {
    return 1;
  }
  return printf("Wrapped %" PRIu64 "\n", wrapped) < 0 ? 1 : 0;
}

File: benchmarks/executable/restaurant_uint_wrapping_shift_left.rs
// Rust 2024 reference for the Restaurant UInt wrapping-shift-left workload.

use std::hint::black_box;
use std::io::{self, Write};

fn wrapping_shift_left_u64(value: u64, count: u64) -> Option<u64> {
    let count = u32::try_from(count).ok()?;
    if count >= u64::BITS {
        return None;
    }
    Some(value.wrapping_shl(count))
}

fn main() {
    let Some(wrapped) = wrapping_shift_left_u64(
        black_box(u64::MAX),
        black_box(1_u64),
    ) else {
        std::process::exit(1);
    };
    let mut stdout = io::stdout().lock();
    if writeln!(stdout, "Wrapped {wrapped}").is_err() {
        std::process::exit(1);
    }
}
